game-core/src/main/java/games/strategy/net/PlayerNameAssigner.java

```diff
@@ -0,0 +1,72 @@
+package games.strategy.net;
+
+import java.net.InetAddress;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+
+/**
+ * Utility class that will assign a name to a newly logging in player.
+ */
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public final class PlayerNameAssigner {
+
+  /**
+   * Name string returned in the case of a hacked client sending us an invalid/unexpected name.
+   */
+  @VisibleForTesting
+  static final String DUMMY_NAME = "aa";
+
+  /**
+   * Returns a node name, based on the specified node name, that is unique across all nodes. The node name is made
+   * unique by adding a numbered suffix to the existing node name. For example, for the second node with the name "foo",
+   * this method will return "foo (1)".
+   *
+   * @param desiredName The name being requested by a new user joining.
+   * @param ipAddress The IP address of the user that is joining. Used to determine if the user
+   *        is already connected under a different name.
+   * @param loggedInNodes Collection of nodes that have already joined.
+   */
+  public static String assignName(
+      final String desiredName,
+      final InetAddress ipAddress,
+      final Collection<INode> loggedInNodes) {
+    Preconditions.checkNotNull(ipAddress);
+    Preconditions.checkNotNull(loggedInNodes);
+
+    String currentName =
+        (desiredName == null || desiredName.length() < 3)
+            ? DUMMY_NAME
+            : findLoggedInName(ipAddress, loggedInNodes).orElse(desiredName);
+    if (currentName.length() > 50) {
+      currentName = currentName.substring(0, 50);
+    }
+    final Collection<String> playerNames = new HashSet<>(
+        loggedInNodes.stream()
+            .map(INode::getName)
+            .parallel()
```

Comment:
Is parallel really increasing performance here?

Reply:
It should, right? Instead of iterating over the list and mapping and adding one at a time, we'll do "n" elements at a time. 

With a 'min' operation, a parallel may not improve performance as the min element needs to be held in memory, but mapping to a set it should be done faster

Reply:
Well in this particular scenario the only thing this is parallelizing is the collection step which is probably not that much faster or even slower because the collections per thread have to be merged somewhere

Reply:
Looks like the answer is 'no' for as-is, if "parallel" is put before the map then it is slightly slower.

I wrote a test script that did a similar map operation over 5M elements and got these results (elapsed time is in milliseconds):

```
Sequential elapsed: 3503, element count: 5000000
Parallel first: 4928, element count: 5000000
Parallel last: 3987, element count: 5000000
```
test script:
````
  private static class NameHolder {
    @Getter
    private final String name = String.valueOf(Math.random());
  }

public static void main(String[] args) {
    final int collectionSize = 5_000_000;
    List<NameHolder> names = new ArrayList<>(collectionSize);
    for(int i = 0; i < collectionSize; i ++) {
      names.add(new NameHolder());
    }

    long start = System.currentTimeMillis();
    final Collection<String> notParallel = new HashSet<>(names.stream()
        .map(NameHolder::getName)
        .collect(Collectors.toSet()));

    long elapsed = System.currentTimeMillis() - start;

    System.out.println("Sequential elapsed: " + elapsed + ", element count: " + notParallel.size());

    start = System.currentTimeMillis();
    final Collection<String> parallelFirst = new HashSet<>(names.stream()
        .parallel()
        .map(NameHolder::getName)
        .collect(Collectors.toSet()));

    elapsed = System.currentTimeMillis() - start;
    System.out.println("Parallel first: " + elapsed + ", element count: " + parallelFirst.size());

    start = System.currentTimeMillis();
    final Collection<String> parallelLast = new HashSet<>(names.stream()
        .map(NameHolder::getName)
        .parallel()
        .collect(Collectors.toSet()));

    elapsed = System.currentTimeMillis() - start;
    System.out.println("Parallel last: " + elapsed + ", element count: " + parallelLast.size());
}
```

Reply:
> if "parallel" is put before the map then it is slightly slower.

I was expected that to be the result, I was wrong there; it's actually slowest when the mapping is done in parallel. That is interesting.

Reply:
I also wonder a bit what happens when the resulting set is not then added back to a new hash set... :thinking:

Reply:
If the results are not added to a hashset, it does get a bit faster (and sequential still wins):
```
Sequential elapsed: 2493, element count: 5000000
Parallel first: 3308, element count: 5000000
Parallel last: 3474, element count: 5000000
```

Reply:
FYI, as stream execution is lazy, it shouldn't matter where `parallel()` is called, as long as the parallel flag is set before invoking the terminal operation (`collect()` in this case).  From the [Javadocs](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/stream/package-summary.html#Parallelism):

> The stream pipeline is executed sequentially or in parallel depending on the mode of the stream on which the terminal operation is invoked. The sequential or parallel mode of a stream can be determined with the `BaseStream.isParallel()` method, and the stream's mode can be modified with the `BaseStream.sequential()` and `BaseStream.parallel()` operations. **The most recent sequential or parallel mode setting applies to the execution of the _entire_ stream pipeline.**